ログイン認証機能を追加

yama998877 opened the pull request:
deviseのログイン認証機能を追加してユーザー一覧ページとユーザー詳細ページを追加し、必須要件を満たすようにカスタマイズしました。

yama998877 committed:
差分を確認しながら直すためにコミット

yama998877 committed:
deviseのカスタマイズ用の辞書ファイルを編集

キーと翻訳文を追加。
変更を加えなかったデフォルトのキーと翻訳文を削除。
生成した辞書ファイルに必要なキーと翻訳文がなかったので追加。

yama998877 committed:
ユーザー認証機能、ユーザーページ、ユーザーリストを実装

郵便番号、住所、自己紹介のデータカラムを追加。
郵便番号のバリデーションを実装。
ログイン、ログアウト、ユーザーデータ更新後の移動先ページを変更。
ユーザー編集ページに郵便番号、住所、自己紹介を追加。
ユーザー詳細ページを追加。

yuuu commented on app/controllers/application_controller.rb:
```diff
@@ -1,4 +1,18 @@
+  end
+
+  # ログアウト後、ログインページに移動
+  def after_sign_out_path_for(_resource)
```
この記載を省略してもログアウト後はログインページに遷移するはずですが、こちらの記載は必要でしょうか？

yama998877 replied:
ご指摘ありがとうございます🙇‍♂️
rootを設定していないのでログアウト後は`/`に移動してしまうので記載しました。
`after_sign_out_path_for(_resource)`を削除して`routes.rb`に`root 'books#index'`を追加すればログインページに遷移しますが

<img width="635" alt="Cursor_と_BooksApp" src="https://github.com/user-attachments/assets/f948e1df-f48c-43b7-b0fb-b2108d081483" />


「ログアウトしました。」の通知が出ずに


<img width="634" alt="BooksApp" src="https://github.com/user-attachments/assets/0b207e80-70f9-4858-ae83-7ecc11777fd3" />

「ログインもしくはアカウント登録してください。」のアラートが出るので必要だと感じました。
rootを設定して`after_sign_out_path_for`を省略した方が良いでしょうか？

yuuu replied:
@yama998877 
こちら、rootを指定していなかったのですね。意図が理解できました 👍 

> rootを設定してafter_sign_out_path_forを省略した方が良いでしょうか？

rootを設定するかどうかはお任せします。
記載の通り、rootを設定してログアウト後にそこに飛ばしても、結局ログイン画面へリダイレクトされてアラートが出てしまうので、 `after_sign_out_path_for` の指定は必要ですね。

rootをLPのようなログイン不要な画面にしていると、デフォルトの挙動がうまく機能するのですが、今回のようなログイン必須のシステムでは指定しておいた方が綺麗です 😃

yuuu commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -0,0 +1,68 @@
+
+  # PUT /resource
+  def update
+    devise_parameter_sanitizer.permit(:account_update, keys: %i[postal_code address self_introduction])
```
このメソッドを呼び出すという方針自体は正しいのですが、こちらのコメントにも記載があるとおり `before_action` で呼び出す方法が一般的です。

https://github.com/heartcombo/devise/blob/fec67f98f26fcd9a79072e4581b1bd40d0c7fa1d/lib/devise/parameter_sanitizer.rb#L21

今の書き方だと、今後 `RegistrationController#update` 以外の箇所でユーザーを更新したくなった際にうまく動かない懸念があるので、上記コメントの呼び方に修正しましょう。

yama998877 replied:
ご指摘ありがとうございます。
`application_ controller.rb`にコードを書き直して、`before_action`でメソッドを呼び出すように修正しました🙇‍♂️

yuuu commented on app/models/user.rb:
```diff
@@ -0,0 +1,10 @@
+  devise :database_authenticatable, :registerable, :recoverable, :rememberable, :validatable
+
+  validates :postal_code, length: { is: 7 }, on: :update, allow_blank: true,
+                          format: { with: /\A[0-9]+\z/, message: 'は半角数字のみが使えます' }
```
エラーメッセージもI18n対応しましょう。
参考: https://railsguides.jp/v7.0/i18n.html#%E3%82%A8%E3%83%A9%E3%83%BC%E3%83%A1%E3%83%83%E3%82%BB%E3%83%BC%E3%82%B8%E3%81%AE%E3%82%B9%E3%82%B3%E3%83%BC%E3%83%97

yama998877 replied:
ご指摘ありがとうございます。
`ja.yml`にキーと訳文を追加してバリデーションエラーメッセージがi18nに対応できるように修正しました🙇‍♂️

yuuu commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,314 @@
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
ここはdeviseから送信されるメールのFromにあたり部分なので、 `no-reply@example.com` のようなそれらしいアドレスに変更しておきましょう。

実際の開発では自分が取得したドメイン名を使用することが多いです。今回はあくまで提出課題なので `example.com` で問題はありません 👍

yama998877 replied:
ご指摘ありがとうございます。
`no-reply@example.com`に変更しました🙏

yuuu commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,314 @@
+
+  # ==> Configuration for :validatable
+  # Range for password length.
+  config.password_length = 6..128
```
昨今のセキュリティ情勢を踏まえると6文字のパスワードは心もとないですね。8文字以上とするのはどうでしょう？

yama998877 replied:
ご指摘ありがとうございます。
８文字以上に変更しました。

yuuu commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,11 @@
+
+<ul>
+  <% @users.each do |user| %>
+  <li><%= link_to "#{user.email}", user %></li>
```
ここの `"#{user.email}"` は `user.email` だけでも同じ結果になりますね 👍

yama998877 replied:
ご指摘ありがとうございます。
`user.email`に修正しました🙏

yuuu commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,31 @@
+  </p>
+
+<nav class="page-nav">
+  <% if current_user.email == @user.email %>
```
確かにメールアドレスはユニークなのでこの方法でも問題はないですが、ここは `if current_user == @user` として問題ないように見えました。

yama998877 replied:
ご指摘ありがとうございます。
デバックして確認してみても問題なかったです。
`current_user == @user`に修正しました🙇‍♂️

yama998877 committed:
RegistrationController#updateのdevise_parameter_sanitizer.permitを削除

application_controller.rbにbefore_actionを使った呼び出し方法に修正するために削除。

yama998877 committed:
configure_permitted_parametersメソッドを追加。

RegistrationController#updateで削除したコードをconfigure_permitted_parametersメソッドとしてbefore_actionで呼び出して利用。

yama998877 committed:
postal_codeの国際化されていないバリデーションエラーメッセージを修正

user.rbから国際化されていないバリデーションエラーメッセージを削除。
ja.ymlにpostal_codeのバリデーションエラーメッセージのキーと訳文を追加。

yama998877 committed:
ユーザーリストのリンクのコードを修正

"#{user.email}"をuser.emailに修正。

yama998877 committed:
詳細ページのリンクのif文の条件式を修正

yama998877 committed:
deviseの設定を変更

再設定メールのアドレスをno-replay@example.comに変更。
パスワードの最低桁数を６文字から８文字に変更。
